cpp/src/parquet/arrow/reader.cc

```diff
@@ -453,229 +483,226 @@ class NestedListReader : public ColumnReaderImpl {
-  Status DefLevelsToNullArray(std::shared_ptr<Buffer>* null_bitmap, int64_t* null_count);
+  const std::shared_ptr<ReaderContext> ctx_;
+  const std::shared_ptr<Field> filtered_field_;
+  const ::parquet::internal::LevelInfo level_info_;
+  const std::vector<std::unique_ptr<ColumnReaderImpl>> children_;
+  ColumnReaderImpl* def_rep_level_child_ = nullptr;
+  bool has_repeated_child_;
 };
 
-Status StructReader::DefLevelsToNullArray(std::shared_ptr<Buffer>* null_bitmap_out,
-                                          int64_t* null_count_out) {
-  auto null_count = 0;
-  const int16_t* def_levels_data;
-  int64_t def_levels_length;
-  RETURN_NOT_OK(GetDefLevels(&def_levels_data, &def_levels_length));
-  ARROW_ASSIGN_OR_RAISE(auto null_bitmap,
-                        AllocateEmptyBitmap(def_levels_length, ctx_->pool));
-  uint8_t* null_bitmap_ptr = null_bitmap->mutable_data();
-  for (int64_t i = 0; i < def_levels_length; i++) {
-    if (def_levels_data[i] < struct_def_level_) {
-      // Mark null
-      null_count += 1;
-    } else {
-      DCHECK_EQ(def_levels_data[i], struct_def_level_);
-      ::arrow::BitUtil::SetBit(null_bitmap_ptr, i);
-    }
-  }
-
-  *null_count_out = null_count;
-  *null_bitmap_out = (null_count == 0) ? nullptr : null_bitmap;
-  return Status::OK();
-}
-
-// TODO(itaiin): Consider caching the results of this calculation -
-//   note that this is only used once for each read for now
 Status StructReader::GetDefLevels(const int16_t** data, int64_t* length) {
   *data = nullptr;
   if (children_.size() == 0) {
-    // Empty struct
     *length = 0;
-    return Status::OK();
+    return Status::Invalid("StructReader had no childre");
   }
 
-  // We have at least one child
-  const int16_t* child_def_levels;
-  int64_t child_length = 0;
-  bool found_nullable_child = false;
-  int16_t* result_levels = nullptr;
-
-  int child_index = 0;
-  while (child_index < static_cast<int>(children_.size())) {
-    if (!children_[child_index]->field()->nullable()) {
-      ++child_index;
-      continue;
-    }
-    RETURN_NOT_OK(children_[child_index]->GetDefLevels(&child_def_levels, &child_length));
-    auto size = child_length * sizeof(int16_t);
-    ARROW_ASSIGN_OR_RAISE(def_levels_buffer_, AllocateResizableBuffer(size, ctx_->pool));
-    // Initialize with the minimal def level
-    std::memset(def_levels_buffer_->mutable_data(), -1, size);
-    result_levels = reinterpret_cast<int16_t*>(def_levels_buffer_->mutable_data());
-    found_nullable_child = true;
-    break;
-  }
+  // This method should only be called when this struct or one of its parents
+  // are optional/repeated or it has a repeated child.
+  // Meaning all children must have rep/def levels associated
+  // with them.
+  RETURN_NOT_OK(def_rep_level_child_->GetDefLevels(data, length));
 
-  if (!found_nullable_child) {
+  if (*data == nullptr) {
+    // Only happens if there are actually 0 rows available.
     *data = nullptr;
```

Comment:
Isn't this redundant with your `if` condition above? Also, why does `length` need to be filled out explicitly below?
(doesn't `def_rep_level_child_->GetDefLevels` do it?).

Reply:
Yes I think it is, removed.